### PR TITLE
feat(inline-notification): add action button with click handler

### LINKED ISSUE #3282: Inline Notification: add secondary action button
### Summary

We'd like to add an "action" button to the `InlineNotification` component, like this:

Screen Shot 2019-07-03 at 3 14 47 PM

### Justification

In our product, there are situations where we want to provide an action to our users when a notification comes up.

### "Must have" functionality

- Button label text prop
- Click handler for the action button

### Specific timeline issues / requests

ASAP 😅 

### Available extra resources

I can help!
I have a suggested way to do this & will submit a PR for feedback shortly.

### LINKED ISSUE #3323: Ghost button: small version is too large (36px height)
## Detailed description

The small ghost `Button` is too large (36px height). It should be 32px tall based on conversation: https://github.com/carbon-design-system/carbon/pull/3283#issuecomment-509766196

See: http://react.carbondesignsystem.com/?path=/story/buttons--default

Screen Shot 2019-07-10 at 9 05 11 AM

### PR DESCRIPTION
**Pull request by jendowns**
Closes #3282 
Closes #3323 

I went ahead and made this PR because we already had an implementation for this button in mind, from our own component library.

I realize I rushed ahead with this so I'm *very* open for feedback about how you want this to look, or even if you don't think it belongs in this library. I can also provide more product examples of us needing to use an action button like this, if you prefer. Just let me know! 👍 

#### Changelog

**New**

- add an action button to the `InlineNotification` component with `actionButtonLabel` and `onActionButtonClick` props
- add style mixin for repeating some styles to be themed (inverse)
- update stories to include knobs for this new button

**Changed**

- add hover color for the close button because it looks like it didn't have one

#### Testing / Reviewing

Can be viewed in the deployment. I recommend switching up the themes to see if you like the colors picked.

### CODE REVIEW

File: packages/components/src/components/notification/_inline-notification.scss
Comment (jendowns) [inline packages/components/src/components/notification/_inline-notification.scss:143]: Since this component is "inverted" and there's not really a good inverted `$link-01` color, this was my approach for inverting theme token colors 🤔

File: packages/components/src/components/notification/_inline-notification.scss
Comment (jendowns) [inline packages/components/src/components/notification/_inline-notification.scss]: Note: I'm not sure if this is the color you want to use? Though it should include something, as right now it doesn't look like there's a hover color at all

**Comment by asudoh** [comment]
You are quick @jendowns! 😅 Let me just repeat what I put in the original issue; I think it'll depend on the design @carbon-design-system/design would give you, but my suggestion is simply adding `actions` prop with `PropTypes.node`. In this way you can add any content you want as the action buttons. (An alternative is `children`, though it's placed in an undesirable place for an action button)

**Comment by jendowns** [comment]
Thanks @asudoh! I made that change just now 👍

**Comment by jeanservaas** [comment]
@jendowns 

our notification design has morphed a bit structurally since chiu's earlier iterations. We actually don't have the hairline separating the close 'x' from the rest of the inline notification anymore. We got rid of it so we could keep the 'x' 16px from the right side of the inline notification and 14px from the top like the type. That way... when the notification goes to two lines and behind, things don't get weird in terms of alignment. Carbon's component doesn't have a hover state on the close icon either... it's just a cursor change.

After talking to @aagonzales and @laurenmrice — we agreed this treatment, using a ghost button for the action button would be more structurally in line with the carbon component. 

image

Also, one other thing to note, because this is a high contrast components use Gray 80... which we haven't rolled out yet, we don't have a token for the ghost button hover state in this scenario.

**Comment by jendowns** [comment]
Thank you @jeanservaas! I removed the borders as you pointed out & made sure the "action" button is a ghost button.

Also I was wondering about the ghost button height compared to the close button? Right now that close button is `3rem`. It looks like the ghost button in those screenshots is smaller?

I just wanted to confirm that you'd like to see the ghost button look shorter versus the close button next to it?

Screen Shot 2019-07-08 at 3 55 19 PM



Thank you!

**Comment by jeanservaas** [comment]
Hi @jendowns !

So right now, if you look at the carbon notification component, we have not hover for close, it's just a cursor change -- so we don't have to worry about matching the action button height to that close button hover square.

The action button height i used was 32px -- we usually use the small buttons inside of components.

image

File: packages/react/src/components/Notification/Notification.js
Comment (asudoh) [inline packages/react/src/components/Notification/Notification.js]: Can we define something like `<NotificationActionButton>` component (probably defined as a named export in this file) so this block can just be `{actions}`?

**Review by asudoh — COMMENTED** [review]
Thank you for your update @jendowns!

**Comment by jendowns** [comment]
Hey @jeanservaas -- thanks for the info! A couple things I noticed that seem different from your screenshot versus what I see in the code right now --

1. It looks like the close icon you have there is bigger than what's live? Right now the close button used has an `Close16` icon (though it looks a bit smaller than that to me? 🤔) Anyway I think that will affect how the button text and close icon "lock up" when side by side

2. The small ghost buttons are unfortunately larger than ordinary ghost buttons! 😲 Small ghost buttons are `36px` tall, versus `32px` for normal ghost buttons. (EDIT: Oh I think it may be the transparent 2px border!)

Screen Shot 2019-07-09 at 10 23 45 AM
Screen Shot 2019-07-09 at 10 23 31 AM

(I'm using the current React storybook as reference for these screenshots: http://react.carbondesignsystem.com/?path=/story/buttons--default)

Anyway what do you think about that? I'd like to avoid making too many style overrides to this ghost button in the inline notification, and I fear I may be getting into some guessing territory. So would you mind taking a look at the screenshots when you get a chance & letting me know if any of the measurements need adjusting?

Thanks!

-------

### EDIT:

One more note about this height difference for these small ghost buttons...

If I force it to be 2rem/32px, without any other adjustments, the text is a bit off the baseline:
Screen Shot 2019-07-09 at 10 36 02 AM

Whereas it looks great if I keep it 2.25rem/36px (ignoring that close button lockup 😅):

Screen Shot 2019-07-09 at 10 37 34 AM

If you are okay with me changing the padding of this ghost button as well... I could adjust it to `height: 2rem` and `padding: 3px 1rem`, then it looks great & very close to your mockup:

Screen Shot 2019-07-09 at 10 39 34 AM

I just want to be sure that it's okay for me to make these types of customizations 😅 Thank you again!

-----

### EDIT 2:

I went ahead made the ghost button smaller (32px) and also adjusted the padding, so that it looks as close as possible to your mockups 👍 Only ended up with the 2 overrides (height and padding)

**Comment by jeanservaas** [comment]
@jendowns thanks for all your work and thoughts on this!

- Good catch; I checked the close 'x' icon and the design kit is pulling a 20px icon there. So we just need to do a little investigating and decide where we landed on that and either modify the code or the kit so they align.

_Update: @laurenmrice confirmed this is a dev bug for modal and notification; we chose 20px icons in those components on purpose. I have to log an issue._

- I just checked with @vpicone about the ghost button issue... looks like ghost button is not subtracting the padding the way the other buttons are. Did you already fix that? Vince also noticed that the field selection under ghost button gives the ghost button hover state the extra right padding that the other buttons have, which is incorrect. We also need to fix that haha. Oh ghost button...

- in terms of this alignment issue... are you suggesting that in order to get the alignment we have to keep the 36px hover state height? Sorry, not sure I understand that one, ideally we could get the 32px hover state height and get alignment, but yes type alignment is the priority there.


> image

**Comment by vpicone** [comment]
@jendowns yeah the padding on ghost buttons doesn't account for the transparent border for some reason? The 3px transparent border is correct and inline with the rest of the buttons. It's kind of significant for when focused.

Also the `field` size still has a 60px padding right 🤕 which looks like a bug.

**Comment by jendowns** [comment]
Thanks for the info @vpicone! Did you want me to make any changes here regarding that?

Also no worries @jeanservaas, my comment was kind of all over the place -- I ended up going with making that ghost button `32px` high & then adding a padding value to it so that it matches your screenshots above. Did you like the way the deployed example looked? I can make adjustments if necessary -- if you prefer, I could even make that icon `Close20` for you in this PR?

@asudoh I also made the changes you asked for above -- could you take another look when you get a chance?

File: packages/react/src/components/Notification/Notification.js
Comment (asudoh) [inline packages/react/src/components/Notification/Notification.js]: Can we define `className`, `children` (instead of `label`) and `onClick` props instead of just `actions`?

File: packages/react/src/components/Notification/Notification.js
Comment (asudoh) [inline packages/react/src/components/Notification/Notification.js]: I meant just `{actions}` here, so user of notification can do `<InlineNotification ... actions={<NotificationActionButton>Action button</NotificationActionButton>}>`

File: packages/react/src/components/Notification/Notification.js
Comment (vpicone) [reply packages/react/src/components/Notification/Notification.js]: Yeah I’m not sure why we’d want to lump all of the props together as a single object. Updating a property on the object wouldn’t force a re-render due to shallow compare. So a dynamic onclick might not function as expected. Just splitting the action up into Its own props for the button is the way to go I think.

File: packages/components/src/components/notification/_inline-notification.scss
Comment (vpicone) [reply packages/components/src/components/notification/_inline-notification.scss:143]: Hmmm I’m not sure this scales really well if we add a
new theme or if someone uses their own theme. I think either making an inverted link token would be appropriate. blue 40 is the inverted color for the dark themes. We could apply that by default then opt for the regular link color in low contrast mode.

File: packages/components/src/components/notification/_inline-notification.scss
Comment (vpicone) [inline packages/components/src/components/notification/_inline-notification.scss]: Do we need the extra small specificity since these are always small? Can we instead fix the small buttons (which should be 32px) and then this button can just have the color overrides?

Comment (jendowns) [reply packages/components/src/components/notification/_inline-notification.scss]: As of right now, it is required to work correctly. I'm waiting for another reviewer to give me the go-ahead to fix something outside of the scope of this PR. The small ghost button issue has been identified but not reported, so I'm working only with what is the reality fo the code.

EDIT: went ahead and reported it here: https://github.com/carbon-design-system/carbon/issues/3323

File: packages/react/src/components/Notification/Notification.js
Comment (jendowns) [reply packages/react/src/components/Notification/Notification.js]: Thanks @asudoh!

Thank you for your additional comment @vpicone -- if you are confused about this usage, please feel to ping me in Slack & I can point you to some examples of similar button series we have in GHE.

File: packages/components/src/components/notification/_inline-notification.scss
Comment (jendowns) [reply packages/components/src/components/notification/_inline-notification.scss:143]: >  I think either making an inverted link token would be appropriate.

What was the "or" here? For "either... or ..." ? 

---

The token approach is fine -- perhaps a general token for inverted link colors like `$inverted-link`. But it's worth noting that there's also the problem of the ghost button hover color. And either way I need some design input here.

@shixiedesign what do you think about a new invert link color token here, so that links appearing on an inverted background ui color work correctly?

EDIT: I mean, `$inverse-link`

Comment (jendowns) [reply packages/components/src/components/notification/_inline-notification.scss:143]: I went ahead and made an issue (https://github.com/carbon-design-system/carbon/issues/3329) and PR (https://github.com/carbon-design-system/carbon/pull/3330) for this new color token 🤔

**Comment by jendowns** [comment]
Thank you again for your feedback @asudoh! I made the corrections you suggested above.

~Though I'm confused about this test failure -- do you know what the issue could be? ...~  OH nevermind! 🕵️‍♀️ I found it! Left a default value of `action: {}` -- should be good now

**Comment by jeanservaas** [comment]
> Did you like the way the deployed example looked? I can make adjustments if necessary -- if you prefer, I could even make that icon Close20 for you in this PR?

Hi @jendowns sorry for the delay, yes, i liked the way your example looked! I was just trying to make sure that the hover state height was 32px, not higher (i.e. is the padding you added transparent?)

Also, yes, it would be great if you could make that close icon 20px!

Thanks! 

ps I think @vpicone would like you to make that change to the `field` size bug on ghost b. too if you have a chance!

**Review by asudoh — APPROVED** [review]
LGTM 👍 once visual review completes - Thank you so much @jendowns!

**Comment by jendowns** [comment]
Hey @jeanservaas -- no worries! This is ready for visual review again, whenever you get a chance.

I was able to make the close button 20px in a separate PR that was merged yesterday, so the change is now reflected here. 

As for the height of the small ghost button here, it is 32px now ~& I'm going to open a PR separately to make sure that small ghost buttons in general are the right height~ 👍 (EDIT: Nevermind about that, I fixed the small ghost buttons here in *this* PR just now, so the height in the `InlineNotification` is correct, as well as small ghost `Buttons` if you look at that `Button` variation directly)

**Review by jeanservaas — APPROVED** [review]
LGTM @jendowns ! Master contributor!

**Comment by jeanservaas** [comment]
Ah, one other thing! I think we could capture this in another PR... but I just noticed the low-contrast variant...

image

I'd think in that instance we'd want the hover to be white... and we'd use the $link-01 token for the ghost button in the low contrast mode (blue 60) 

image

**Also 2: TOKEN DRAMA @aagonzales @laurenmrice** 
We don't have tokens for these at the moment:

- the hover state on the gray 80 high contrast notification background
- the white hover state on the color tints of the low contrast notifications
- the blue 40 link color for the ghost button on the high contrast notifications... in a dark theme $link-01 would pull that value, but the high contrast or dark notifications are being used in light themes... so $link-01 would pull blue 60

**Comment by jendowns** [comment]
Thanks @jeanservaas! So I added a couple low contrast styles that you mentioned... 🤔 Let me know what you think!

- ghost button text color: `$link-01`
- hover background: `$ui-background`

Screen Shot 2019-07-11 at 1 33 39 PM
Screen Shot 2019-07-11 at 1 33 45 PM

One thing that looks a little off to me is the `active` state 🤔
**What color do you think it should it be?**

Screen Shot 2019-07-11 at 1 34 11 PM

**Comment by jeanservaas** [comment]
@jendowns 

GOOD CATCH... man notifiCATIONS!!!! they're killin us.

Just kicked around these two ideas with @laurenmrice and @aagonzales and wanted to get your opinion. Design could live with both... but we liked the idea of making the text darker in this scenario v. using the gray... it's just a little bit more of an exception.

**Option 1 (Design preferred)**

image

**Option 2**

image

**Comment by jendowns** [comment]
Thanks @jeanservaas! I applied option 1 & you should be able to see it in the preview now. 🎉

**Review by laurenmrice — APPROVED** [review]
Looks great Jen! Thanks so much !

**Comment by vpicone** [comment]
@jendowns Hey! Thanks so much for getting this in! One thing that concerns me in hindsight is if someone with a screen reader would have enough time to interact with an action. Wondering if you explored this at all and if we should open an issue to investigate now that this is merged.

**Comment by jendowns** [comment]
Hey @vpicone I don’t remember how long this type of notification stays up in our application — sometimes a given notification linger depending on the situation. It’s a very good question though & sounds like it would be more related to a11y testing within a given application’s implementation (where they would set the time outs etc) versus here in the library where the component is just “there” & doesn’t disappear.

Though there could still be an opportunity to write up some guidance on a11y and notifications in general? 🤔

**Comment by elizabethsjudd** [comment]
Toast notifications are a tricky one with a11y because they typically auto-hide but need to meet the requirements of this checkpoint: https://www.ibm.com/able/guidelines/ci162/pause_stop_hide.html. I'd be really curious to get you're perspective on how this could be addressed in offerings. (right now we are trying to limit the number of toast notifications used at all in WH)

Inline notifications shouldn't be an issue as they are typically persistent.

**Pull request merged**
